### ISSUE TITLE
Help requested: Simple VAE in Gen

### ISSUE THREAD

**Issue by innuo** [probcomp/Gen.jl#302]
First, thanks for all your great work in building this excellent prob-prog framework. I have been trying to understand how to use Gen to build a a Variational Autoencoder but haven't succeeded in getting it to work. Any pointers would be greatly appreciated.

In the following example I am trying to learn the 4 params of the generative model and the 5 params of the variational model using blackbox VI.  I've played with the gradient descent learning rates, trying only to update a few of the params in the 'update' call, but I've not been able make black_box_vi produce anything but NANs. 

```
@gen function gen_model(n) #Hypothesized model (Decoder)
	@param c::Float64
	@param lin::Float64
        @param quad::Float64
        @param cub::Float64
	
	for i in 1:n
		z = @trace(normal(0, 1), (:z, i))
		y = @trace(normal(c + lin * z + quad * z^2 + cub * z^3, 1) , (:y, i))
	end
   	return n
end;

@gen function var_model(y) #variational model (Encoder), also a simple polynomial model for the mean
       @param c::Float64
       @param lin::Float64
       @param quad::Float64
       @param cub::Float64
       @param log_std::Float64
	
	for i in 1:length(y)
		@trace(normal(c + lin * y[i] + quad * y[i]^2 + cub * y[i]^3, exp(log_std)),(:z, i))
	end
	return 
end

init_param!(gen_model, :c, 0.)
init_param!(gen_model, :lin, 0.)
init_param!(gen_model, :quad, 0.)
init_param!(gen_model, :cub, 0.)

init_param!(var_model, :c, 0.)
init_param!(var_model, :lin, 0.)
init_param!(var_model, :quad, 0.)
init_param!(var_model, :cub, 0.)
init_param!(var_model, :log_std, 0.)

function vi_training(data)
	observations = choicemap()
	
	for (i, y) in enumerate(data)
            observations[(:y, i)] = y
        end
	
	update = ParamUpdate(GradientDescent(0.00001, 1000), 
		gen_model => [:c, :lin, :quad, :cub],
		var_model => [:c, :lin, :quad, :cub, :log_std])
	
	black_box_vi!(gen_model,  (length(data),), 
		observations, var_model, (data,), update;
		iters=20, samples_per_iter=50, verbose=true)

	return (c0=get_param(gen_model, :c), c1=get_param(gen_model, :lin),
		    c2=get_param(gen_model, :quad), c3=get_param(gen_model, :cub))
	
end


n = 1000
latents = [normal(0, 1) for _ in 1:n]
data = 0.1 .+ 0.01*latents .+ .3*latents.^3 #generate some `data`

vi_training(data)
```

**Comment by innuo** [comment]
Any help with this issue would be much appreciated. If any part of the question is unclear, I would glad to provide clarifications.

**Comment by innuo** [comment]
Here is what the training loop produces.

```
julia> vi_training(data)
iter 1; est objective: -1729.3194718932612
iter 2; est objective: -4.686529374563846e6
iter 3; est objective: -6.54884322065037e14
iter 4; est objective: NaN
iter 5; est objective: NaN
...
iter 20; est objective: NaN
(c0 = 0.0, c1 = 0.0, c2 = 0.0, c3 = 0.0)
```
What am I doing wrong?

**Comment by ztangent** [comment]
Hey @innuo, sorry you've been facing this issue. I've yet to try building a VAE in Gen myself, but it's possible that this is due to the overflow bug when calculating the `logpdf` of a normal distribution, as described in #321. Hopefully this will no longer be an issue once we merge in #321, but in the meantime, is it possible that some of the samples are leading to very large values of `std`, or very large values of `mu - y`, where `y` is the observed value?

**Comment by alex-lew** [comment]
Hi @innuo,

Thanks for trying out Gen! Sorry for the delay in getting back to you on this.

One thing to note about this code is that `black_box_vi` on its own assumes the model parameters are fixed, and only updates the parameters of the variational family, `var_model`. You'll need to use `accumulate_param_gradients!` on the model, and then `apply!` to update the parameters. (You can look at the body of `black_box_vi!` to see how this is done for `var_model`.

@marcoct -- maybe we should make a version of `black_box_vi!` that also updates model parameters?

One other thing, @innuo, is that Gen currently doesn't support variance reduction strategies like the reparameterization trick, so even though VAEs can be implemented, it will be with higher-variance gradient estimates than you'd get if you implemented the algorithm by hand. @marcoct has worked out some changes that will fix this, and add support for various variance reduction strategies, but I'm not sure when they'll be implemented.

All best,
Alex

**Comment by marcoct** [comment]
@innuo 

As a first step of debugging, I would check that optimizing the variational family succesfully converges with fixed model parameters. (If it doesn't, then you can e.g. reduce the step size until it does). I suspect you're seeing a gradient diverge because step sizes are too large.

After optimization of the variational family alone is working, then you can add a for loop that interleaves calls to `black_box_vi!` with `iters=1` with a few lines of code that estimate the gradients for the model and update its parameters. The resulting code would look something like this:
```julia

..

model_update = ParamUpdate(GradientDescent(0.00001, 1000), 
		gen_model => [:c, :lin, :quad, :cub])

var_update = ParamUpdate(GradientDescent(0.00001, 1000),
		var_model => [:c, :lin, :quad, :cub, :log_std])
	
for iter in 1:max_iter
    
    # update variational family
    (_, var_traces, _) = black_box_vi!(gen_model,  (length(data),), 
		observations, var_model, (data,), update;
		iters=1, samples_per_iter=50, verbose=true)

    # update model
    for var_trace in var_traces
        constraints = merge(observations, get_choices(var_trace))
        (model_trace, _) = generate(gen_model,  (length(data),), constraints)
        accumulate_param_gradients!(model_trace, nothing)
    end
    apply!(model_update)
   
end
```

To understand this code it will be useful to understand the implementation of `black_box_vi!` first: https://github.com/probcomp/Gen.jl/blob/cd49151234ea4d7b97adc6f22cc0e666379c8432/src/inference/variational.jl#L101-L134

**Comment by marcoct** [comment]
@alex-lew Good suggestion about allowing `black_box_vi!` to optionally update the model as well!

**Comment by marcoct** [comment]
@innuo While we do not have the reparametrization trick supported right now in Gen's inference library, we do have an implementation of variational inference with Monte Carlo objectives (VIMCO), which uses multiple samples from the variational family to tighten the ELBO, while also leveraging the multiple samples to reduce gradient variance. After you have the `black_box_vi!` version working on a very small test case, it should be possible to swap out `black_box_vi!` for `black_box_vimco!`, which has the same interface.

**Comment by innuo** [comment]
Thanks for all the suggestions! It looks like I should spend some time trying to understand the implementation of the black_box_vi.  

However, interestingly, I was unable to get the update to work even if I initialized the model parameters to their true values. Perhaps the variance of the grad estimates for the var_model were still too large for my chosen step size.

**Comment by marcoct** [comment]
Note that I just edited my code sample above: The traces returned by black_box_vi! are traces of the variational family, not the generative model, so we have to construct the model traces from the variational family traces when doing the update of the generative model. (It would be a small code change to our library function to pass back the model traces as well so this work isn't duplicated; or better yet, as Alex suggested, just add the optional generative model parameter update here as well).

**Comment by alex-lew** [comment]
@innuo @marcoct Re: numerically stable gradient updates, it might be helpful to use a more sophisticated optimizer, like ADAM, as well. The [GenFluxOptimizers.jl](https://github.com/probcomp/GenFluxOptimizers.jl) package will let you swap out GradientDescent with ADAM or another optimizer.

**Comment by marcoct** [comment]
Here is a draft PR that implements @alex-lew's suggestion to add the update to the generative model to these functions: https://github.com/probcomp/Gen.jl/pull/333

@innuo This PR might not me merged immediately, but the commit should provide useful context for how to implement a VAE in Gen.

**Comment by marcoct** [comment]
@innuo I added a unit test to the PR, which shows a [mimimal implementation of a toy VAE in Gen](https://github.com/probcomp/Gen.jl/pull/333/commits/2cd26f0f36e7d6e534bcc09579e3a50aea29fbc8) that uses `black_box_vi!` to do training of the model as well as the inference model (and similarly for `black_box_vimco!`). There are various other ways of setting things up, but this is a minimal example using the built-in library. Note that I used  that example as a test case so that (i) the true posterior was within the variational family, and (ii) it was possible to compute the true marginal likelihood and the optimal model parameter that maximizes it.
